presto-native-execution/presto_cpp/main/PrestoServer.cpp

```diff
@@ -1378,6 +1388,57 @@ static protocol::Duration getUptime(
   return protocol::Duration(seconds, protocol::TimeUnit::SECONDS);
 }
 
+void PrestoServer::convertToVeloxPlan(
+    proxygen::HTTPMessage* message,
+    proxygen::ResponseHandler* downstream,
+    const std::vector<std::unique_ptr<folly::IOBuf>>& body) {
+  protocol::ExecutionFailureInfo failure;
+  try {
+    auto headers = message->getHeaders();
+
+    std::ostringstream oss;
+    for (auto& buf : body) {
+      oss << std::string((const char*)buf->data(), buf->length());
+    }
+    std::string planFragmentJson = oss.str();
+    protocol::PlanFragment planFragment = json::parse(planFragmentJson);
+
+    auto queryCtx = core::QueryCtx::create();
+    VeloxInteractiveQueryPlanConverter converter(queryCtx.get(), pool_.get());
+
+    // Create static taskId and empty TableWriteInfo needed for plan conversion
+    protocol::TaskId taskId = "velox-plan-conversion.0.0.0";
+    auto tableWriteInfo = std::make_shared<protocol::TableWriteInfo>();
+
+    // Attempt to convert the plan fragment to a Velox plan
+    if (auto writeNode =
+            std::dynamic_pointer_cast<const protocol::TableWriterNode>(
+                planFragment.root)) {
+      // TableWriteInfo is not yet built at the planning stage, so we can not
+      // fully convert a TableWriteNode and skip that node of the fragment.
+      converter.toVeloxQueryPlan(writeNode->source, tableWriteInfo, taskId);
```

Comment:
@BryanCutler @tdcmeehan : A design point to keep in mind is the extent to which this code might catch errors. 

For the most part the Presto -> Velox plan conversion would catch many cases. 

But there is also a category of errors that happen when the operators are constructed from the Velox PlanNode(s). e.g. Scalar, aggregate, window function objects get created only on the Project/HashAgg//Window operator initialization. So certain function validations happen only then. 

Take a look at https://github.com/facebookincubator/velox/blob/main/velox/exec/LocalPlanner.cpp#L416.

As we gain more experience running this code, its likely we will want to abstract some part of the Driver creation as well for the plan validation.

Reply:
We knew that this wouldn't catch all errors, but hopefully catch most early enough during queueing to be useful. It will be great to categorize what native errors are not caught by this and see what else can be done about those.